File: Binary_Search/34_Find_First_and_Last_Position_of_Element_in_Sorted_Array/CPP/34_Find_First_and_Last_Position_of_Element_in_Sorted_Array.cpp
// Problem link : https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> ans;
        ans.push_back(firstOcc(nums,target));
        ans.push_back(lastOcc(nums,target));
        return ans;
    }
    
    int lastOcc(vector<int> &nums,int target)
    {
        int low=0, high = nums.size()-1, mid, ans = -1;
        
        while(low <= high)
        {
            mid = (low + high)/2;
            
            if(nums[mid] < target)
            {
                low = mid + 1;
            }
            else if(nums[mid] > target)
            {
                high = mid - 1;
            }
            else
            {
                low = mid + 1;
                ans = mid;
            }
        }
        
        return ans;
    }
    
    int firstOcc(vector<int> &nums,int target)
    {
        int low=0, high = nums.size()-1, mid, ans = -1;
        
        while(low <= high)
        {
            mid = (low + high)/2;
            
            if(nums[mid] < target)
            {
                low = mid + 1;
            }
            else if(nums[mid] > target)
            {
                high = mid - 1;
            }
            else
            {
                high = mid - 1;
                ans = mid;
            }
        }
        
        return ans;
    }
};

File: Binary_Search/34_Find_First_and_Last_Position_of_Element_in_Sorted_Array/JAVA/34_Find_First_and_Last_Position_of_Element_in_Sorted_Array.java
// Problem link : https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/

class Solution {
    public int[] searchRange(int[] nums, int target) {
        int [] ans = new int[2];
        
        ans[0] = find_first_last_occ(nums, target, 1);
        ans[1] = find_first_last_occ(nums, target, 0);
        
        return ans;
    }
    
    int find_first_last_occ(int[] nums, int target, int flag)
    {
        int low = 0, high = nums.length-1, mid, ans = -1;
        
        while(low <= high)
        {
            mid = (low + high)/2;
            
            if(target > nums[mid])
                low = mid + 1;
            else if(target < nums[mid])
                high = mid - 1;
            else
            {
                if(flag == 0)
                    low = mid + 1;
                else
                    high = mid - 1;
                
                ans = mid;
            }
        }
        
        return ans;
    }
}
